src/irmin/tree.ml

```diff
@@ -924,6 +924,18 @@ module Make (P : Backend.S) = struct
       | Error _ as e -> e
       | Ok m -> Ok (StepMap.bindings m)
 
+    let seq_of_updates update_map value_bindings =
+      (* This operation can be costly for a large update_map. *)
+      let value_bindings =
+        Seq.filter (fun (s, _) -> not (StepMap.mem s update_map)) value_bindings
+      in
+      let update_seq =
+        StepMap.to_seq update_map
+        |> Seq.filter_map (fun (s, elt) ->
+               match elt with Remove -> None | Add e -> Some (s, e))
+      in
+      Seq.append value_bindings update_seq
```

Comment:
Perhaps we could go with something like this, which I think is a bit more order-preserving:

```suggestion
    (* Given an unordered sequence of bindings and an update map, return a new
       sequence with the updates applied, preserving the original order of steps. *)
    let rec seq_of_updates updates value_bindings () =
      if StepMap.is_empty updates then
        (* Short-circuit return if we have no more updates to apply *)
        value_bindings ()
      else
        match value_bindings () with
        | Seq.Nil ->
            (* Any remaining [adds] introduce new steps (not part of the
               original order), which we add at the end. *)
            let remainder =
              StepMap.to_seq updates
              |> Seq.filter_map (fun (s, elt) ->
                     match elt with Remove -> None | Add e -> Some (s, e))
            in
            remainder ()
        | Seq.Cons (((step, _) as old_binding), value_bindings) -> (
            match StepMap.find_opt step updates with
            | None ->
                Seq.Cons (old_binding, seq_of_updates updates value_bindings)
            | Some update -> (
                let updates = StepMap.remove step updates in
                let tail = seq_of_updates updates value_bindings in
                match update with
                | Add v' -> Seq.Cons ((step, v'), tail)
                | Remove -> tail ()))
```